chain/ethereum/tx.go

```diff
@@ -0,0 +1,110 @@
+package ethereum
+
+import (
+	"context"
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	types2 "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/renproject/multichain/api/account"
+	"github.com/renproject/multichain/api/address"
+	"github.com/renproject/multichain/api/contract"
+	"github.com/renproject/pack"
+	"math/big"
+)
+
+// TxBuilder represents a transaction builder that builds transactions to be
+// broadcasted to the ethereum network. The TxBuilder is configured using a
+// gas price and gas limit.
+type TxBuilder struct {
+	ChainID     *big.Int
+	BlockNumber *big.Int
+}
+
+// NewTxBuilder creates a new transaction builder.
+func NewTxBuilder(chainID, blockNumber *big.Int) TxBuilder {
+	return TxBuilder{chainID, blockNumber}
+}
+
+// BuildTx receives transaction fields and constructs a new transaction.
+func (txBuilder TxBuilder) BuildTx(ctx context.Context, from, to address.Address, value, nonce, gasLimit, gasPrice, gasCap pack.U256, payload pack.Bytes) (account.Tx, error) {
+	toAddr, err := NewAddressFromHex(string(pack.String(to)))
+	if err != nil {
+		return nil, fmt.Errorf("bad to address '%v': %v", to, err)
+	}
+	return &Tx{
+		types2.NewTransaction(nonce.Int().Uint64(), common.Address(toAddr), value.Int(), gasLimit.Int().Uint64(), gasPrice.Int(), payload),
+		types2.MakeSigner(&params.ChainConfig{
+			ChainID: txBuilder.ChainID,
+		}, txBuilder.BlockNumber),
+	}, nil
+}
+
+// Tx represents a ethereum transaction, encapsulating a payload/data and its
+// signer.
+type Tx struct {
+	ethTx  *types2.Transaction
+	signer types2.Signer
+}
+
+// Hash returns the hash that uniquely identifies the transaction.
+// Generally, hashes are irreversible hash functions that consume the
+// content of the transaction.
+func (tx Tx) Hash() pack.Bytes {
+	return tx.ethTx.Hash().Bytes()
+}
+
+// From returns the address that is sending the transaction. Generally,
+// this is also the address that must sign the transaction.
+func (tx Tx) From() address.Address {
+	return address.Address("")
+}
```

Comment:
You can use [this](https://github.com/ethereum/go-ethereum/blob/v1.10.3/core/types/transaction_signing.go#L126) to get the sender of the transaction if there is a valid signature attached to the transaction object. Only if the `tx` is unsigned you would return an empty address.